pkg/foxtrot/api.go

```diff
@@ -0,0 +1,95 @@
+package foxtrot
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strconv"
+
+	"foxygo.at/s/errs"
+	"foxygo.at/s/httpe"
+)
+
+// api is a REST inspired HTTP API for accessing foxtrot data,
+// registration and authentication
+//
+// /api/login POST
+// /api/register POST
+// /api/history?room=NAME[&before=MESSAGE_ID|TIMESTAMP&count=N]
+//
+// Not yet implemented:
+// /api/user/NAME/
+// /api/user/NAME/avatar
+// /api/room/NAME # create new room.
+type api struct {
+	db   *db
+	auth *authenticator
+}
+
+func newAPI(db *db, auth *authenticator) *api {
+	a := &api{
+		db:   db,
+		auth: auth,
+	}
+	return a
+}
+
+func (a *api) wireRoutes(basePath string, mux *http.ServeMux) {
+	mux.Handle(basePath+"/login", httpe.Must(httpe.Post, a.login))
+	mux.Handle(basePath+"/register", httpe.Must(httpe.Post, a.register))
+	mux.Handle(basePath+"/history", httpe.Must(httpe.Get, a.history))
+}
+
+type creds struct {
+	Name     string `json:"name"`
+	Password string `json:"password"`
+}
+
+func (a *api) login(w http.ResponseWriter, r *http.Request) error {
+	c := creds{}
+	defer r.Body.Close() //nolint: errcheck
+	if err := json.NewDecoder(r.Body).Decode(&c); err != nil {
+		return errs.New(httpe.ErrBadRequest, err)
+	}
+	u, err := a.auth.login(r.Context(), c.Name, c.Password)
+	if err != nil {
+		w.Header().Set("WWW-Authenticate", `Bearer realm="Write access to foxtrot chat"`)
+		return httpe.ErrUnauthorized
+	}
+	return json.NewEncoder(w).Encode(u)
+}
+
+func (a *api) register(w http.ResponseWriter, r *http.Request) error {
+	c := creds{}
+	defer r.Body.Close() //nolint: errcheck
+	if err := json.NewDecoder(r.Body).Decode(&c); err != nil {
+		return errs.New(httpe.ErrBadRequest, err)
+	}
+	u := User{Name: c.Name}
+	if err := a.auth.register(r.Context(), &u, c.Password); err != nil {
+		if errors.Is(err, errDBDuplicate) {
+			return httpe.ErrBadRequest
```

Comment:
I wonder whether this should return a different result or something - how would an API user differentiate this error from the bad json decoding one? This could be considered a user "error", as the user input cannot be used. The other would be a client error. I'm not sure what it would return otherwise, unless we json encoded error codes too.

Perhaps just leave it for now, but something to keep in mind.

Reply:
yeah, agreed, 4xx should probably have a response body.